File: src/syntax/FStar.Syntax.Cache.fs
(*
   Copyright 2008-2018 Microsoft Research

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*)
#light "off"
module FStar.Syntax.Cache

open FStar.ST
open FStar.All
(* Type definitions for the core AST *)

(* Prims is used for bootstrapping *)
open Prims
open FStar
open FStar.Util
open FStar.Range
open FStar.Ident
open FStar.Dyn

module S = FStar.Syntax.Syntax

type cache<'a> = imap<(S.term * 'a)>

let create (_:unit) = imap_create 100

let add (c:cache<'a>) (t:S.term) (v:'a) : unit =
  let hash = generic_hash t in
  imap_add c hash (t, v)

//an optional comparator function for additional checking on a cache hit
let find (c:cache<'a>) (t:S.term) (f_opt:option<(S.term -> S.term -> bool)>) : option<'a> =
  t |> generic_hash |> imap_try_find c |> (fun v_opt ->
    match v_opt with
    | None         -> None
    | Some (t', v) ->
      (match f_opt with
       | None   -> Some v
       | Some f -> if f t t' then Some v else None))

File: src/ocaml-output/FStar_Syntax_Cache.ml
open Prims
type 'a cache = (FStar_Syntax_Syntax.term * 'a) FStar_Util.imap
let create : 'Auu____15 . unit -> 'Auu____15 FStar_Util.imap =
  fun uu____22  -> FStar_Util.imap_create (Prims.parse_int "100") 
let add : 'a . 'a cache -> FStar_Syntax_Syntax.term -> 'a -> unit =
  fun c  ->
    fun t  ->
      fun v1  ->
        let hash = FStar_Util.generic_hash t  in
        FStar_Util.imap_add c hash (t, v1)
  
let find :
  'a .
    'a cache ->
      FStar_Syntax_Syntax.term ->
        (FStar_Syntax_Syntax.term -> FStar_Syntax_Syntax.term -> Prims.bool)
          FStar_Pervasives_Native.option -> 'a FStar_Pervasives_Native.option
  =
  fun c  ->
    fun t  ->
      fun f_opt  ->
        let uu____114 =
          let uu____121 = FStar_All.pipe_right t FStar_Util.generic_hash  in
          FStar_All.pipe_right uu____121 (FStar_Util.imap_try_find c)  in
        FStar_All.pipe_right uu____114
          (fun v_opt  ->
             match v_opt with
             | FStar_Pervasives_Native.None  -> FStar_Pervasives_Native.None
             | FStar_Pervasives_Native.Some (t',v1) ->
                 (match f_opt with
                  | FStar_Pervasives_Native.None  ->
                      FStar_Pervasives_Native.Some v1
                  | FStar_Pervasives_Native.Some f ->
                      let uu____189 = f t t'  in
                      if uu____189
                      then FStar_Pervasives_Native.Some v1
                      else FStar_Pervasives_Native.None))
  